GCTC-NTGC/gc-digital-talent
# Avoid network errors in ProfilePage stories

Closes: https://github.com/GCTC-NTGC/gc-digital-talent/issues/2885

This is an alternative approach to the suggestion of refactoring props, which @tristan-orourke and I discussed on Thurs. Putting it here for consideration :)

Generally, a perk of using hooks in react is that they are often backed by a context that can almost always be injected with fake state or methods. If hooks are used, this means no special prop architecture or practices are required.

I'm submitted this as something to merge and have in our repertoire, not as an immediate replacement for any practices we're currently using. The hope is that if the two approaches are permitted to coexist for a bit, we'll end up choosing the one that serves us best!

With this decorator, if there's any graphql hook for queries, for example a `getMyThings` operation, we can just mock out the minimal response in stories. (This is comparable to how urql docs recommend)

```js
// MyComponent.stories.tsx
export default {
  // ...
  parameters: {
    apiResponses: {
      getMyThings: {
        data: [
          {name: 'thing1'},
          {name: 'thing2'},
        ]
      }
    }
  }
} as Meta;`

// ...

// Or can set on story itself
export const MyStory = Template.bind({});
MyStory.parameters = {
  apiResponses: {
    getMyThings: {
      data: [
        {name: 'thing3-override'},
      ]
    }
  }
}

```

## To Do
- [x] Create a decorator for mocking API responses
- [x] Add decorator a minimal API response to component's stories.
- [x] make decorator global and available by default
- [x] more descriptive story names
- [x] convert decorator to typescript

## To Test
1. Visit built storybook via the link in PR checks
2. Search "Profile Form": https://61e099a1bb1465003a73d3ce-rzqbwfilyw.chromatic.com/?path=/story/profile-form--completed-without-experiences
3. Confirm that component stories now renders without error.

---

This PR is in the spirit of urql's own recommended way of working with storybook, though I didn't use their addon:
https://formidable.com/open-source/urql/docs/advanced/testing/#handling-multiple-hooks
https://storybook.js.org/addons/@urql/storybook-addon

(The above is not required reading, just to show that mocking queries is recommended over refactoring data into props.)
